stan/math/prim/prob/discrete_range_lcdf.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef STAN_MATH_PRIM_PROB_DISCRETE_RANGE_LCDF_HPP
+#define STAN_MATH_PRIM_PROB_DISCRETE_RANGE_LCDF_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/inv.hpp>
+#include <stan/math/prim/fun/log.hpp>
+#include <stan/math/prim/fun/max_size.hpp>
+#include <stan/math/prim/fun/size_zero.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/** \ingroup prob_dists
+ * Return the log CDF of a discrete range distribution for the given y,
+ * lower and upper bounds (all integers).
+ *
+ * `y`, `lower` and `upper` can each be a scalar or a one-dimensional container.
+ * Any container arguments must be the same size.
+ *
+ * @tparam T_y type of scalar, either int or std::vector<int>
+ * @tparam T_lower type of lower bound, either int or std::vector<int>
+ * @tparam T_upper type of upper bound, either int or std::vector<int>
+ *
+ * @param y integer random variable
+ * @param lower integer lower bound
+ * @param upper integer upper bound
+ * @return The log CDF evaluated at the specified arguments. If containers are
+ * supplied, returns the sum of the log CDFs.
+ * @throw std::domain_error if upper is smaller than lower.
+ * @throw std::invalid_argument if non-scalar arguments are of different
+ * sizes.
+ */
+template <typename T_y, typename T_lower, typename T_upper>
+double discrete_range_lcdf(const T_y& y, const T_lower& lower,
+                           const T_upper& upper) {
+  static const char* function = "discrete_range_lcdf";
+
+  if (size_zero(y, lower, upper)) {
+    return 0.0;
+  }
+
+  check_not_nan(function, "Random variable", y);
+  check_consistent_sizes(function, "Lower bound parameter", lower,
+                         "Upper bound parameter", upper);
+  check_greater_or_equal(function, "Upper bound parameter", upper, lower);
+
+  scalar_seq_view<T_y> y_vec(y);
+  scalar_seq_view<T_lower> lower_vec(lower);
+  scalar_seq_view<T_upper> upper_vec(upper);
+  size_t size_y_lower = max_size(y, lower);
+  size_t N = max_size(y, lower, upper);
+
+  for (size_t n = 0; n < N; ++n) {
+    const double y_dbl = value_of(y_vec[n]);
+    if (y_dbl < value_of(lower_vec[n])) {
```

Comment:
These are inconsistent.  Both this and the next conditional should return `LOG_ZERO` if anythng, but really they should be throwing exceptions because `y` is out of scope.  Also, all this should be int, not double.

Reply:
Following what is done in other lcdfs, if `y` is below the lower bound, the lcdf returns `LOG_ZERO`, but if it's above the upper boound it returns `log(1)`, by definition of a log CDF. But I'll start throwing, so this is moot.